widgets: Support markdown in widgets.

roanster007 opened the pull request:
Previously, markdown was not supported in widgets like polls and todo. This is fixed by adding an inline only backend processor ZulipInlineMarkdown, and using it to render the poll options or tasks and descriptions of todo before adding them to the database.

- Commt 1 (`fbb12bc`): Add `ZulipInlineMarkdown` processor to render inline only patterns.
- Commit 2 (`e4ceba6`): Support inline markdown in polls.
- Commit 3 (`1197b5f`): Support inline markdown in todo widgets.  

Fixes: #21917

[CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Support.20links.20in.20Polls/near/1674037)

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="548" alt="image" src="https://github.com/zulip/zulip/assets/97049524/c0770cb6-630c-46db-be62-43f29d3381ac">

<img width="544" alt="image" src="https://github.com/zulip/zulip/assets/97049524/66084967-cb55-4036-82c6-715f17bbbfbf">



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

bupd commented:
@roanster007 Reviewed the changes locally <time and basic markdown seems to be working fine.

![poll-time](https://github.com/zulip/zulip/assets/89722848/8685bcc9-b1b2-45f4-88fe-f334aa46216e)

I've noticed that the current process of adding time to a poll in Zulip involves typing the time in the compose box and then copying it to the "Add Option" field. While this works well, it would be even more user-friendly to directly implement time rendering for additional options.

I suggest extending the time rendering functionality to the "Add Option" field, allowing users to input time in a similar manner. This enhancement would streamline the process and provide a consistent experience across the poll creation interface.

Implementation Details:

When users type a time in the "Add Option" field, the system should recognize and render it accordingly.
Ensure that the time rendering follows the same format as when entered in the compose box.
Benefits:

Improved user experience by eliminating the need to copy-paste time from the compose box to the "Add Option" field.
Consistency in time rendering across different input fields.
I believe this enhancement would make the poll creation process even more efficient and user-friendly. I'd be happy to contribute to the implementation if needed.

The decision should be made by project contributors to include the above suggestion or not.

bupd reviewed (approved):
It works.

alya commented:
Thanks! For polls, links and emoji look good to me, but the time should be rendered like it is in messages. Compare:

![Screenshot 2024-01-04 at 10 54 22 AM](https://github.com/zulip/zulip/assets/2090066/bd7f02f1-959a-48b9-8961-1be9809e92b6)

alya commented:
For todo lists:

1. The elements should work in the description part of the item as well, not just the first part.
2. These changes break the crossed out formatting for checked items.

![Screenshot 2024-01-04 at 11 00 04 AM](https://github.com/zulip/zulip/assets/2090066/e5827b0f-cec7-4d1d-9e1a-7b03baa8af51)

alya commented:
@bupd I agree that we should ideally have a better way to input times (and emoji), but let's make sure we get the rendering part of this working first.

roanster007 commented:
@alya I have updated the PR and screenshots as per the feedback.

alya commented:
Thanks! It looks like the space after the `:` and before the descriptions might have disappeared?

alya commented:
Other than that, the screenshots look good to me. You can add the "maintainer review" label (via Zulipbot) and ping @N-Shar-ma to take a look once that's been fixed.

roanster007 commented:
@zulipbot add "maintainer review"

roanster007 commented:
@N-Shar-ma could you have a look at this..

N-Shar-ma commented on web/src/poll_widget.ts:
```diff
@@ -225,10 +228,9 @@ export function activate({
         const html = render_widgets_poll_widget_results(widget_data);
         $elem.find("ul.poll-widget").html(html);
-
+        render_poll_timestamp($elem);
```
Directly pass `$elem.find("ul.poll-widget")` instead to this function. That will make it more generalised, so also reusable by todo widgets, eliminating the near duplicate function in `todo_widget.js`

N-Shar-ma replied:
Would be even better if you could extract this as a function in `rendered_markdown` where this code is from, and call that here. 

We always want to minimize code duplication as much as possible

roanster007 replied:
I felt using the `update_elements` of `rendered_markdown` would be unreasonable since that method would try to look out for lot of other elements apart form `time` in `$elem` like mentions, codehilite,... but those for sure won't be rendered inside the widget by the inline processor.

Moreover, importing `rendered_markdown.js` method into `poll_widget.ts` is causing `Could not find a declaration file for module './rendered_markdown'. 'rendered_markdown.js' implicitly has an 'any' type`

N-Shar-ma replied:
I meant extracting just the relevant code as a new function and calling that, not the pre existing whole `update_elements` function.

But yes, the non typescript file import into this typescript file still remains an issue :thinking:

N-Shar-ma commented on web/styles/widgets.css:
```diff
@@ -166,7 +181,7 @@
         color: hsl(0deg 0% 45%);
         padding-left: 4px;
-        padding-top: 28px;
+        padding-top: 2px;
```
This can be increased to 3 or 4px... It currently looks a little off to me:
![image](https://github.com/zulip/zulip/assets/68962290/28f5c1b3-ea5c-438f-b32f-98b2514b75b6)

N-Shar-ma replied:
Actually you might have to rethink the html structure here... For a multiline option the voters' names take up an entire column of space:
![image](https://github.com/zulip/zulip/assets/68962290/8152d759-840c-4298-b25a-e81dddaee942)

roanster007 replied:
How about now?
<img width="543" alt="image" src="https://github.com/zulip/zulip/assets/97049524/cec8e954-6063-4e34-9f06-dab47e43b733">

N-Shar-ma replied:
Hmm, the 1st line being spaced extra from the rest due to the vote count button looks a little odd... I think we want to keep that button in its own column, but keep the option and voter names inline. Would this be possible?

roanster007 replied:
yeah it is possible, I'll update it... I actually changed it as I thought you wanted something similar to the present one.

N-Shar-ma commented on web/templates/widgets/todo_widget_tasks.hbs:
```diff
@@ -8,9 +8,9 @@
             <div>
                 {{#if completed}}
-                <strike><strong>{{ task }}</strong>{{#if desc }}: {{ desc }}{{/if}}</strike>
+                <strike><strong class="task_complete">{{{ task }}}</strong>{{#if desc }}<span class="task_desc_complete"">: </span><span class="description_complete"">{{{ desc }}}</span>{{/if}}</strike>
```
There's an extra `"` after the `task_desc_complete` class... Also, can you explain why these new classes are needed?

N-Shar-ma replied:
Extra `"` after `description_complete` class too

roanster007 replied:
Oh sorry... I missed it. So the reason for these new classes was -- `task`, `:` and `description` are now rendered as separate elements. So without them, sometimes the `:` is not staying inline, while other times, all three of them go to the next line. So I thought I would enclose them within tags to make them inline separately.

N-Shar-ma reviewed (changes requested):
Reviewed the frontend code and have left some comments. 

Someone else who is familiar with the backend markdown processor will be much better suited to review those parts; sorry I've not worked on that part of the codebase so far

roanster007 committed:
markdown: Create an inline markdown processor.

This commit creates an inline only markdown processor, that
renders inline patterns.

roanster007 committed:
polls: Render markdown in polls.

This commit allows polls to render inline patterns using the
inline markdown processor.

Fixes #21917.

roanster007 committed:
todo: Render markdown in todo widgets.

This commit allows todo to render inline patterns using the
inline markdown processor.

nimishmedatwal commented:
I believe this pr also fixes #21674 #16821 #12947 and #21674.

haxfn commented:
Images, requested in [#user questions > images in polls](https://chat.zulip.org/#narrow/channel/138-user-questions/topic/images.20in.20polls/with/2420997)
